### PR TITLE
change uid-map to spec-ref-map

### PR DESCRIPTION
**Pull request by Waschi97**
i would like to work on this PR

### CODE REVIEW

File: src/topp/QualityControl.cpp [hunk 1]
```diff
@@ -269,7 +279,7 @@ class TOPPQualityControl : public TOPPBase
         const auto& ptr_to_map_entry = map_to_identifier.find(unique_run_path);
         if (ptr_to_map_entry == map_to_identifier.end())
         {
-          OPENMS_LOG_ERROR << "FeatureXML (MS run '" << unique_run_path << "') does not correspond to ConsensusXML (run not found). Check input!\n";
+          OPENMS_LOG_ERROR << "FeatureXML (MS run '" << unique_run_path[0] << "') does not correspond to ConsensusXML (run not found). Check input!\n";
```
Comment (cbielow) [inline src/topp/QualityControl.cpp]: you can use `ListUtils::concatenate(x, ", ")` to show the whole StringList

File: src/topp/QualityControl.cpp [hunk 2]
```diff
@@ -385,18 +394,38 @@ class TOPPQualityControl : public TOPPBase
...
     {
       // for empty PIs which were created by a metric
       if (ref_pep_id.getHits().empty()) continue;
 
-      if (!ref_pep_id.metaValueExists("UID")) // PepID doesn't has ID, needs to have MetaValue
+      PeptideIdentification pep_id;
+      const ProteinIdentification& prot_id = *(identifier_to_protID.at(ref_pep_id.getIdentifier()));
+      StringList filenames;
+      prot_id.getPrimaryMSRunPath(filenames);
+      if (filenames.empty())
```
Comment (cbielow) [inline src/topp/QualityControl.cpp]: check for empty can be done once and earlier (when creating the `identifier_to_prot` map)

File: src/topp/QualityControl.cpp [hunk 3]
```diff
@@ -385,18 +394,38 @@ class TOPPQualityControl : public TOPPBase
...
+      {
+        throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No MS run path annotated at ProteinIdentification.");
+      }
+      if (filenames.size() == 1)
       {
-        throw(Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No unique ID at peptideidentifications found. Please run PeptideIndexer with '-addUID'.\n"));
+        String UID = filenames[0] + ref_pep_id.getMetaValue("spectrum_reference").toString();
+        pep_id = *customID_to_pepID.at(UID);
+      }
+      else
```
Comment (cbielow) [inline src/topp/QualityControl.cpp:446]: use `else if ...`

File: src/topp/QualityControl.cpp [hunk 4]
```diff
@@ -406,16 +435,38 @@ class TOPPQualityControl : public TOPPBase
...
+      if (filenames.empty())
+      {
+        throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No MS run path annotated at ProteinIdentification.");
+      }
+      if (filenames.size() == 1)
+      {
+        pep_id.setMetaValue("cf_id", group_id);
+        String UID = filenames[0] + pep_id.getMetaValue("spectrum_reference").toString();
+        customID_to_pepID[UID] = &pep_id;
+        continue;
```
Comment (cbielow) [inline src/topp/QualityControl.cpp]: remove the the `continue` and use `else if `

File: src/topp/QualityControl.cpp [hunk 5]
```diff
@@ -406,16 +435,38 @@ class TOPPQualityControl : public TOPPBase
...
+      }
+      if (pep_id.metaValueExists("map_index"))
+      {
+        pep_id.setMetaValue("cf_id", group_id);
```
Comment (cbielow) [inline src/topp/QualityControl.cpp]: this is duplicated in both if-blocks. Move it outside.

**Review by cbielow — CHANGES_REQUESTED** [review]
nice!

if you want, we could add some verfication code here https://github.com/cbielow/OpenMS/blob/develop/src/topp/QualityControl.cpp#L303
do make sure the mapping targeted all PepIDs in the consensusMap, i.e. all PepIds should carry the metavalue "missed_cleavages" or something which we can guarantee to be present. We can also add it later though

**Comment by Waschi97** [comment]
> nice!
> 
> if you want, we could add some verfication code here https://github.com/cbielow/OpenMS/blob/develop/src/topp/QualityControl.cpp#L303
> do make sure the mapping targeted all PepIDs in the consensusMap, i.e. all PepIds should carry the metavalue "missed_cleavages" or something which we can guarantee to be present. We can also add it later though

what do you mean by "verification code"? i think i'm not familiar with that..

**Pull request merged**